kratos/factories/base_factory.h

```diff
@@ -0,0 +1,270 @@
+        const std::string& r_name = Settings["name"].GetString();
+        KRATOS_ERROR_IF_NOT(Has( r_name )) << "Trying to construct a class with type name= " << r_name << "\n" <<
+                                            "Which does not exist. The list of available options (for currently loaded applications) are: \n" <<
+                                            KratosComponents<ClassType>() << std::endl;
+        const ClassType& aux = KratosComponents<ClassType>::Get( r_name );
+        return aux.Create(std::forward<TArgumentsType>(Arguments)...);
+    }
+
+    ///@}
+    ///@name Access
+    ///@{
+
+    ///@}
+    ///@name Inquiry
+    ///@{
+
+    ///@}
+    ///@name Input and output
+    ///@{
+
+    /// Turn back information as a string.
+    std::string Info() const override
+    {
+        return "BaseFactory";
+    }
+
+    /// Print information about this object.
+    void PrintInfo(std::ostream& rOStream) const override
+    {
+        rOStream << this->Info() << std::endl;
+        const auto factory_components = KratosComponents<ClassType>::GetComponents();
+        for (const auto& r_comp : factory_components) {
+            rOStream << "\t" << r_comp.first << std::endl;
+        }
+    }
+
+    /// Print object's data.
+    virtual void PrintData(std::ostream& rOStream) const override
+    {
+//         rOStream << this->Info() << std::endl;
+//         const auto factory_components = KratosComponents<ClassType>::GetComponents();
+//         for (const auto& r_comp : factory_components) {
+//             rOStream << "\t" << r_comp.first << std::endl;
+//         }
+    }
+
+    ///@}
+};
+
+///@}
+///@name Type Definitions
+///@{
+
+///@}
+///@name Input and output
+///@{
+
+/// output stream function
+template<class TClass, typename TAuxiliarClass>
+inline std::ostream& operator << (std::ostream& rOStream,
+                                  const BaseFactory<TClass, TAuxiliarClass>& rThis)
+{
+    rThis.PrintInfo(rOStream);
+    rOStream << std::endl;
+    rThis.PrintData(rOStream);
+
+    return rOStream;
+}
+
+///@}
+
+void RegisterExplicitStrategies();
+void RegisterExplicitBuilders();
+
+typedef ExplicitSolvingStrategy<SparseSpaceType, LocalSpaceType> ExplicitSolvingStrategyType;
+typedef ExplicitBuilder<SparseSpaceType, LocalSpaceType> ExplicitBuilderType;
+
+KRATOS_API_EXTERN template class KRATOS_API(KRATOS_CORE) KratosComponents<ExplicitSolvingStrategyType>;
+
+#ifdef KRATOS_REGISTER_EXPLICIT_STRATEGY
+#undef KRATOS_REGISTER_EXPLICIT_STRATEGY
+#endif
+#define KRATOS_REGISTER_EXPLICIT_STRATEGY(name, reference) \
+    KratosComponents<ExplicitSolvingStrategyType>::Add(name, reference);
+
+KRATOS_API_EXTERN template class KRATOS_API(KRATOS_CORE) KratosComponents<ExplicitBuilderType>;
+
+#ifdef KRATOS_REGISTER_EXPLICIT_BUILDER
+#undef KRATOS_REGISTER_EXPLICIT_BUILDER
+#endif
+#define KRATOS_REGISTER_EXPLICIT_BUILDER(name, reference) \
+    KratosComponents<ExplicitBuilderType>::Add(name, reference);
```

Comment:
can you please move this to a separate file?

this is a baseclass header, those includes would be everywhere otherwise

Reply:
Huuum, this should be in a header file anyway, why not here?

Reply:
because this is a baseclass file which will be included in many places
=> i.e. reducing dependencies is the keywords